### ISSUE TITLE
Export linker

### ISSUE THREAD

**Issue by axic** [ethereum/solc-js#4]
The library linking option is not available in Javascript. Would be very useful to have it in the form of:

`solc.linkBytecode(bytecode, <class name / address pairs>)`
`solc.link(<output JSON from compile>, <class name to link>, <class name / address pairs>)`

This could also obsolete the linking code in [browser-solidity](https://github.com/chriseth/browser-solidity/blob/gh-pages/assets/js/universal-dapp.js#L462) (and perhaps in [truffle](https://github.com/ConsenSys/truffle/blob/master/lib/contracts.js#L353)).

It could be
a) implemented in Javascript in `solc-js` or
b) `solidity/solc/jsonCompiler.cpp` could be extended to export `link()` too (as it is present internally).

**Comment by chriseth** [comment]
This concerns more solidity in general, but I would like to bring this up here:

Solidity does not have something like a "global namespace" on "globally unique namespace hierarchy". This could be a good thing, at least it is again a parallel to javascript (at least npm and I think also es6) where imports only modify the current scope. When seen from "the outside", this creates problems: There is no way to uniquely refer to a contract or library by name, you always have to specify the scope.

If there are two libraries with the same name, how do we uniquely distinguish them for the sakes of the linker? `<filename>/libraryname`?

**Comment by axic** [comment]
Good point.  Not totally sold whether the `<filename>` part should be the a) caller's filename (the file where the class referencing it is) or b) a reference to the import directive.  In the case of option b) a separator other than `/` might be more appropriate.

**Comment by chriseth** [comment]
With `<filename>` I meant the (potentially absolute) path of the file that contains the library definition.

**Comment by redsquirrel** [comment]
Not sure if it makes sense to bring it up here, but one way to resolve naming collisions would be to somehow delegate that conflict to one or more of the emerging Solidity package managers, who would ensure unique names within their registries.

**Comment by axic** [comment]
It is related to import statements too because the library might be defined via a URL (see your Github URL parser in browser-solidity)

**Comment by tcoulter** [comment]
> If there are two libraries with the same name, how do we uniquely distinguish them for the sakes of the linker? <filename>/libraryname?

Can you have two libraries/contracts of the same name within one Solidity file? i.e., 

```
import "SomeLib.sol";
import "./must/go/deeper/SomeLib.sol"; 
```

Assuming both define a Library called `SomeLib`, I assume the second import will overwrite first? If so, this means it's not the linker's problem - there's only one library to link. 

Now, in the case where you used the alternative `import` syntax, I assume there would be two libraries to be linked:

```
import "SomeLib.sol";
import "./must/go/deeper/SomeLib.sol" as DifferentLib;
```

**Comment by axic** [comment]
@tcoulter I think it referred to the point that you can supply multiple Solidity files to the compiler at the same time, which might not be related (sharing the same import chain)?

**Comment by chriseth** [comment]
@tcoulter yes, this is how you deal with name conflicts internally, but still this does not create an externally visible name hierarchy, because the library can be imported into files in various ways. I think the only way to disambiguate libraries is by prefixing the file name where they are defined.

**Comment by chriseth** [comment]
@axic I think a proper way to do this would be to expose things like the "LinkerObject" to the javascript world. The current way of putting the library name in the hex bytecode is just a very crude way of doing it because it is limited to 32 characters. The compiler should expose objects that contain unlinked bytecode and a list of libraries to be linked in. As discussed, these libraries should not only be identified by their name but also by the path to the file they are defined in.

**Comment by axic** [comment]
I think as long as `compile` is allowed to output the crude way of library address placeholders, a crude linker should be also provided.

Do you have a link to the "LinkerObject" somewhere (maybe a C++ header file)?

**Comment by chriseth** [comment]
Sure, the natively compiled `solc` also offers both options. Please take a look at `solidity/libevmasm/LinkerObject.h`.

**Comment by redsquirrel** [comment]
@chriseth:
> As discussed, these libraries should not only be identified by their name but also by the path to the file they are defined in.

It would be great if we could not assume that the source was at a specific file path, allowing for the flexility to dynamically provide the source as a string.

**Comment by tcoulter** [comment]
I agree with @redsquirrel. I don't necessarily understand your comment @chriseth about the externally visible name hierarchy, but I would love to see code namespaces make their way into the linker rather than assuming a file path, i.e.:

This code:

```
import "SomeLib.sol";
import "./must/go/deeper/SomeLib.sol" as DifferentLib;

contract MyContract { ... }
```

might create the following binary:

```
0xabcd1234__SomeLib__________abcd1234abcd1234__DifferentLib__________238947239432"
```

**Comment by chriseth** [comment]
@redsquirrel passing source by string will of course still be possible. In fact, the "driver" will pass a collection of source files as a key-value mapping and the key will be the "file path".

@tcoulter note that `import "SomeLib.sol";` will import everything in the file `"SomeLib.sol"`, i.e. there could be multiple libraries there. Furthermore, through inheritance or by calling internal library functions, `MyContract` drag more dependencies in, which might be out of its context. So in general, I think that naming libraries by how they are imported will not work. Another simple reason for global and against local naming is that it is much easier to link multiple uses of a single library, you only need to search for one name. Finally, the linker is part of the "driver" which knows the external context better than the internal context (because it was that component that supplied the files in the first place).

**Comment by axic** [comment]
@chriseth I was thinking about this and perhaps it would be wiser moving off from putting text into the output. Unless you are fine with having longer than 40 character patterns.

How about using a keccak hash: `____{the first 18 bytes of the hash}` (Uses 4 leading underscores to not break compatibility with the current scheme.)

Additionally the compiler would output the different import statements (paths) in the resulting JSON so the version of them could be determined from the outside (i.e. by a package manager, such as live-libs).

**Comment by chriseth** [comment]
@axic putting the text into the output was only a way to keep some kind of backwards compatibility. This should never ever be used by any serious solidity programmer. We should really provide access to the linker objects from javascript.

**Comment by axic** [comment]
This was implemented via #8. I suggest to close this issue and open a new one when the new version of linking (with metadata and linkerobject) comes around.
